internal/config/sshSetup.go

```diff
@@ -0,0 +1,177 @@
+package config
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"golang.org/x/crypto/ssh"
+)
+
+func createKeyPair(privateKeyPath, publicKeyPath string) error {
+	privateKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return err
+	}
+
+	// Generate and write private key as PEM
+	privateKeyFile, err := os.Create(privateKeyPath)
+	if err != nil {
+		return err
+	}
+	defer privateKeyFile.Close()
+
+	privateKeyPEM := &pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(privateKey),
+	}
+
+	if err := pem.Encode(privateKeyFile, privateKeyPEM); err != nil {
+		return err
+	}
+
+	// Set permissions to private key
+	if err = os.Chmod(privateKeyPath, 0400); err != nil {
+		return err
+	}
+
+	// Generate and write public key
+	pub, err := ssh.NewPublicKey(&privateKey.PublicKey)
+	if err != nil {
+		return err
+	}
+
+	err = ioutil.WriteFile(publicKeyPath, ssh.MarshalAuthorizedKey(pub), 0400)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func knownHost(host string, port int) bool {
+	var output []byte
+	var err error
+
+	if port != 22 {
+		output, err = exec.Command("ssh-keygen", "-H", "-F", fmt.Sprintf("%s:%d", host, port)).CombinedOutput()
+	} else {
+		output, err = exec.Command("ssh-keygen", "-H", "-F", host).CombinedOutput()
+	}
+
+	if err.Error() == "exit status 1" && len(output) > 0 {
+		return true
+	}
```

Comment:
I don't understand this part. The host exists (in `known_hosts`) when it returns an error of `exit status 1` and there is some output on stdout/stderr?

I found an issue with this. When a host has the port number along with it, in the `known_hosts` file it is written as `[ip_address]:port` instead of `ip_address:port`. This leads to it trying to transfer keys again.
For example, with `ssh-keygen -H -F 192.168.0.0:8022` I get an empty output whereas with `ssh-keygen -H -F "[192.168.0.0]:8022"` it gives signature of the node. (IP address change ofc xD)

Making this change gives a nil pointer derefrence though.
```
		output, err = exec.Command("ssh-keygen", "-H", "-F", fmt.Sprintf("[%s]:%d", host, port)).CombinedOutput()
```

Changing the condition to `err == nil` works. Could you confirm this?
```suggestion
	if port != 22 {
		output, err = exec.Command("ssh-keygen", "-H", "-F", fmt.Sprintf("[%s]:%d", host, port)).CombinedOutput()
	} else {
		output, err = exec.Command("ssh-keygen", "-H", "-F", host).CombinedOutput()
	}

	if err == nil && len(output) > 0 {
		return true
	}
```

---

An alternative would be to remove this known hosts check entirely xD. `ssh-copy-id` will not copy the key if it already exists:
```
/usr/bin/ssh-copy-id: WARNING: All keys were skipped because they already exist on the remote system.
```
So it should be harmless to try to transfer keys again. Also, calling commands this way can be rather finicky, especially when their outputs have to be parsed. I'd like to avoid it if possible.

Reply:
Okay I think I'll remove the host check. It is rather weird. Plus since ssh-copy-id handles the case when keys exist, should make life easier (also one less external dependency!).